File: src/main/java/org/nest/codegeneration/sympy/DeltaSolutionTransformer.java
/*
 * Copyright (c) 2015 RWTH Aachen. All rights reserved.
 *
 * http://www.se-rwth.de/
 */
package org.nest.codegeneration.sympy;

import org.nest.nestml._ast.ASTFunctionCall;
import org.nest.nestml._ast.ASTNeuron;
import org.nest.nestml._symboltable.predefined.PredefinedFunctions;
import org.nest.utils.AstUtils;

import java.util.List;

import static java.util.stream.Collectors.toList;
import static org.nest.codegeneration.sympy.AstCreator.createDeclaration;

/**
 * Takes SymPy result with the exact solution of the Delta-shaped PSC and integrates it into the neuron instead of
 * the 'integrate'-call.
 * Produces an altered AST with the the exact solution.
 *
 * @author plotnikov
 */
class DeltaSolutionTransformer extends TransformerBase {

  ASTNeuron addExactSolution(
      final SolverOutput solverOutput, final ASTNeuron astNeuron) {

    ASTNeuron workingVersion = astNeuron;
    //addVariableToInternals(astNeuron, p30File);
    workingVersion.addToInternalBlock(createDeclaration("__h ms = resolution()"));
    workingVersion = addVariableToInternals(workingVersion, solverOutput.const_input);
    workingVersion = addVariableToInternals(workingVersion, solverOutput.ode_var_factor);

    final List<ASTFunctionCall> i_sumCalls = AstUtils.getAll(astNeuron.findEquationsBlock().get(), ASTFunctionCall.class)
        .stream()
        .filter(astFunctionCall -> astFunctionCall.getCalleeName().equals(PredefinedFunctions.CURR_SUM))
        .collect(toList());

    // Apply spikes from the buffer to the state variable
    for (ASTFunctionCall i_sum_call : i_sumCalls) {
      final String bufferName = AstUtils.toString(i_sum_call.getArgs().get(1));
      solverOutput.ode_var_update_instructions.add(astNeuron.getEquations().get(0).getLhs().getName() + "+=" + bufferName);
    }

    workingVersion = replaceIntegrateCallThroughPropagation(workingVersion, solverOutput.ode_var_update_instructions);
    return workingVersion;
  }

}

File: pynestml/solver/DeltaSolutionTransformer.py
#
# DeltaSolutionTransformer.py
#
# This file is part of NEST.
#
# Copyright (C) 2004 The NEST Initiative
#
# NEST is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# NEST is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
from pynestml.solver.SolverOutput import SolverOutput
from pynestml.solver.TransformerBase import TransformerBase
from pynestml.modelprocessor.ASTNeuron import ASTNeuron
from pynestml.modelprocessor.ASTFunctionCall import ASTFunctionCall
from pynestml.modelprocessor.PredefinedFunctions import PredefinedFunctions
from pynestml.modelprocessor.ModelParser import ModelParser
from pynestml.utils.ASTUtils import ASTUtils
from pynestml.codegeneration.ExpressionsPrettyPrinter import ExpressionsPrettyPrinter


class DeltaSolutionTransformer(TransformerBase):
    """
    This class contains a set of methods as used to add solutions to a handed over neuron.
    """

    @classmethod
    def addExactSolution(cls, _solverOutput=None, _neuron=None):
        """
        Adds the exact solution as computed by the solver to the neuron.
        :param _solverOutput: a single solver output object
        :type _solverOutput: SolverOutput
        :param _neuron: a single neuron
        :type _neuron: ASTNeuron
        :return: the modified neuron
        :rtype: ASTNeuron
        """
        assert (_solverOutput is not None and isinstance(_solverOutput, SolverOutput)), \
            '(PyNestML.Solver.DeltaSolution) No or wrong type of output provided (%s)!' % _solverOutput
        assert (_neuron is not None and isinstance(_neuron, ASTNeuron)), \
            '(PyNestML.Solver.DeltaSolution) No or wrong type of neuron provided (%s)!' % _neuron
        _neuron.addToInternalBlock(ModelParser.parseDeclaration('__h ms = resolution()'))
        TransformerBase.addVariableToInternals(_neuron, _solverOutput.const_input)
        TransformerBase.addVariableToInternals(_neuron, _solverOutput.ode_var_factor)
        i_sum_calls = [func for func in ASTUtils.getAll(_neuron.getEquationsBlocks(), ASTFunctionCall)
                       if func.getName() == PredefinedFunctions.CURR_SUM]
        expression_printer = ExpressionsPrettyPrinter()
        # now apply spikes from the buffer to the state variables
        for call in i_sum_calls:
            buffer_name = expression_printer.printExpression(call.getArgs()[1])
            _solverOutput.ode_var_update_instructions.append(
                _neuron.getEquations()[0].getLhs().getName() + '+=' + buffer_name)

        _neuron = TransformerBase.replaceIntegrateCallThroughPropagation(_neuron, _solverOutput.const_input,
                                                                         _solverOutput.ode_var_update_instructions)
        return _neuron
